Replace balenalib base images documentation with Official Docker images (#3180)
balena-io/docs: pages/reference/base-images/balena-base-images.md

 - Fedora official images: [https://hub.docker.com/_/fedora](https://hub.docker.com/_/fedora)
 - Alpine official images: [https://hub.docker.com/_/alpine](https://hub.docker.com/_/alpine)

### Device Architecture Note

Dockerhub uses a different naming convention for device architecture than balena. Use the table below to convert a balena specified architecture to the equivalent architecture listed on Dockerhub. To see the architecture for balena-supported devices, check out the [balena Machine names and architecture page](https://docs.balena.io/reference/base-images/devicetypes/).

As an example, if you are using a Raspberry Pi Zero 2W, the balena machine name page shows the “balena_arch” as “aarch64”. In the table below, aarch64 is referred to as  “arm64v8 (ARMv8 64-bit)” on Dockerhub. Look out for arm64v8 images to find images compatible with the Raspberry Pi Zero 2W device type. 

| Docker Architecture | Balena Architecture (BALENA\_ARCH) | Alternative Names | Example Device Types |
| :---- | :---- | :---- | :---- |
| arm32v6 (ARMv6 32-bit) | rpi | armv6, armhf (in some contexts) | Raspberry Pi 1 model B+, Raspberry Pi Zero/W |

> | arm32v7 (ARMv7 32-bit) | armv7hf | armv7, armv7l, armhf | Raspberry Pi 2, Raspberry Pi 3 model B+ |

Reviewer: I wonder if it might be better to suggest arm64v8 for Raspberry Pi 3 Model B+ ?

Reviewer 2: The list for popular supported 32 bit devices is very slim and this could be a good addition.

Author: I updated the table to have some varied hardware examples.

> | arm64v8 (ARMv8 64-bit) | aarch64 | arm64, armv8 | Raspberry Pi 3 64-bit, Raspberry Pi 4, NVIDIA Jetson |

Reviewer: Also Raspberry Pi 5 and Raspberry Pi Zero 2 W

Reviewer 2: The list is meant to be for reference purposes only. 
@alanb128 I might further recommend not using all examples as RPi and add from different manufacturers too if available.

Reviewer: That's fair :+1:

---
title: Base images
excerpt: Tips for choosing a base image for your project

---

# Choosing a base image

## Where is balenalib?

> Balenalib was the home for 26,000+ IoT-focused Docker images built specifically for balenaCloud and balenaOS. Balena ceased publishing updates to balenalib images on May 23, 2025, however, all of the previously published images are still available and still work. Read [our announcement blog post](https://blog.balena.io/deprecate-balenalib-images/) for more information. 

Reviewer 2: ```suggestion
Balenalib was the home for 26,000+ IoT-focused Docker images built specifically for balenaCloud and balenaOS. Balena stopped publishing updates to balenalib images on May 23, 2025, however, all of the previously published images are still available, and applications using them would still work. Read [our announcement blog post](https://blog.balena.io/deprecate-balenalib-images/) for more information. 
```

> Note: if you are looking for the overview of balena base images (deprecated) that used to be on this page, you can find that here.

Reviewer 2: Use the __note__ notation to render a note box on the docs
Link missing 
```suggestion
Note: The docs for balenalib base images (deprecated) that used to be on this page have been moved. Navigate to PAGE_NAME to find them.
```

> When balena started publishing these images back in 2016, the Docker ecosystem was quite new, and support for ARM 6/7/8 images was almost non-existent. Today, however,  Docker images for all architectures and software stacks are  well supported and maintained. Consequently, this shift meant that balenalib base images' value and advantage have diminished over the past few years.

Reviewer 2: ```suggestion
When balena started publishing these base images back in 2016, the Docker ecosystem was quite new, and support for ARM 6/7/8 images was almost non-existent. Today, however,  Docker images for all architectures and software stacks are well supported and maintained. Consequently, this shift in the larger ecosystem meant that balenalib base images' value and advantage have diminished over the past few years.
```

## How to Pick a Base Image

For general use cases, you can usually start with a standard Linux distribution’s Official Docker Image, and install further requirements and dependencies in your Dockerfile. For instance, here is an example of installing the Nano text editor in the [official Debian image](https://hub.docker.com/_/debian):

```Dockerfile
FROM debian:bookworm-20250407  # This is the official image

# Here we install a utility not included in the base image:

RUN apt update && apt install -y --no-install-recommends nano  

CMD ["sleep", "infinity"]
```
Scroll past the tag list to find more helpful information in the “Overview” tab on the Docker Hub repository for these popular OS images. That information includes supported architectures, image descriptions, and the available variants and tags:
 
 - Debian official images: [https://hub.docker.com/_/debian](https://hub.docker.com/_/debian)
 - Ubuntu official images: [https://hub.docker.com/_/ubuntu/](https://hub.docker.com/_/ubuntu/)
 - Fedora official images: [https://hub.docker.com/_/fedora](https://hub.docker.com/_/fedora)
 - Alpine official images: [https://hub.docker.com/_/alpine](https://hub.docker.com/_/alpine)

> ### Device Architecture Note

Reviewer 2: ```suggestion
### Picking the right base Image by device architecture
```

> Dockerhub uses a different naming convention for device architecture than balena. Use the table below to convert a balena specified architecture to the equivalent architecture listed on Dockerhub. To see the architecture for balena-supported devices, check out the [balena Machine names and architecture page](https://docs.balena.io/reference/base-images/devicetypes/).

Reviewer 2: ```suggestion
Dockerhub uses a different naming convention for device architecture than balena. Use the table below to convert a balena-specified architecture to the equivalent architecture listed on Dockerhub. To see the architecture for balena-supported devices, check out the [balena Machine names and architecture page](/reference/base-images/devicetypes/).
```

this prevent a new page load on the server and improving user experience.

 - Ubuntu official images: [https://hub.docker.com/_/ubuntu/](https://hub.docker.com/_/ubuntu/)
 - Fedora official images: [https://hub.docker.com/_/fedora](https://hub.docker.com/_/fedora)
 - Alpine official images: [https://hub.docker.com/_/alpine](https://hub.docker.com/_/alpine)

### Choose the right device architecture

Dockerhub uses a different naming convention for device architecture than balena. Use the table below to convert a balena specified architecture to the equivalent architecture listed on Dockerhub. To see the architecture for balena-supported devices, check out the [balena Machine names and architecture page](/reference/base-images/devicetypes/).

As an example, if you are using a Raspberry Pi Zero 2W, the balena machine name page shows the “balena_arch” as “aarch64”. In the table below, aarch64 is referred to as  “arm64v8 (ARMv8 64-bit)” on Dockerhub. Look out for arm64v8 images to find images compatible with the Raspberry Pi Zero 2W device type. 

| Docker Architecture | Balena Architecture (BALENA\_ARCH) | Alternative Names | Example Device Types |
| :---- | :---- | :---- | :---- |

> | arm32v6 (ARMv6 32-bit) | rpi | armv6, armhf (in some contexts) | Raspberry Pi 1 model B+, Raspberry Pi Zero/W |

Reviewer: Could you put a capital 'M' on "model" please, to match the phrasing at https://www.raspberrypi.com/products/raspberry-pi-1-model-b-plus/ ?

Author: Sure!

The instructions for installing a language stack from the official binaries is beyond the scope of this document, but many guides exist online. Here are a few examples:

 - Installing Python from source: [https://realpython.com/installing-python/#linux-how-to-build-python-from-source-code](https://realpython.com/installing-python/#linux-how-to-build-python-from-source-code)
 - Installing Node from source: [https://github.com/nodejs/node/blob/main/BUILDING.md](https://github.com/nodejs/node/blob/main/BUILDING.md)
 - Installing Golang from source: [https://linuxhit.com/how-to-install-go-on-linux-using-official-binary-releases/](https://linuxhit.com/how-to-install-go-on-linux-using-official-binary-releases/)
 - Installing OpenJDK from source: [https://openjdk.org/groups/build/doc/building.html](https://openjdk.org/groups/build/doc/building.html)
And so on

Note that these instructions are not necessarily written for building a container, so you may have to adapt the steps to work in a Dockerfile. For example, if an instruction says to “download a file with wget https://go.dev/dl/go1.24.3.linux-amd64.tar.gz" the line in the Dockerfile would be: 

> `RUN https://go.dev/dl/go1.24.3.linux-amd64.tar.gz`

Author: Yes, thanks good catch!
